### ISSUE TITLE
[SUGGESTION] When running tests, don't rebuild if nothing has changed

### ISSUE THREAD

**Issue by sebwills** [microsoft/vscode-dotnettools#1365]
### Describe the feature you'd like

My C# project takes 30s-1m to build, even if there are no changes since the last build.

Every time I run tests using the C# DevKit test runner (i.e. from the Test Results window or Testing tab), the project is rebuilt, even if I have not touched any source files. This adds 30s-1m to my test iteration time.

I would like an option which does not run the build if none of the source files are newer than the target assembly. I realise this could potentially mean not picking up updated nuget package, etc, but that would be an acceptable trade-off. I could use an explicit build command to rebuild if necessary.

### Alternatives considered

I tried adding a wrapper script around the build command that checks for changes. I can get this to work for a `type: process` build task, but not for a `type: dotnet` build task, due to #1363.

But I cannot use this approach for the build invoked by the devkit test runner, because I cannot get the test runner to use my customized build task for the build it invokes (#1364).

In any case, these would be quite involved workarounds involving custom build script wrappers. A simple built-in way for DevKit to short-circuit unnecessary builds when running tests would be a much better UX.

### Environment Information

- Windows and Linux (remote dev)
- VSCode 1.90.2
- C# DevKit 1.9.55

**Comment by nojaf** [comment]
Hello @sebwills, would the experimental flags mentioned in [this comment](https://github.com/microsoft/vscode-dotnettools/issues/1343#issuecomment-2284149112) be helpful?

**Comment by sebwills** [comment]
Hi @nojaf, yes that looks like that would allow me to avoid unnecessary builds - great news!

But, I think it would be much nicer if there was option to rebuild automatically _but only if there have been changes to source_. I can imagine such a setting could be a reasonable default behaviour in future (and aligns with behavior expected of build systems and IDEs), whereas I can't imagine `dotnet.testWindow.disableBuildOnRun` would ever be set as a default because of the confusion resulting from changes not getting picked up in tests by users not aware of the setting.

So, I'm delighted to hear about the option, but I don't think it resolves this suggestion.

**Comment by Plasma** [comment]
Customizing this task would be helpful in order to pass an additional parameter to msbuild to signal its a test project build (we can then reduce the code files being built for faster iterations)

**Comment by sebwills** [comment]
Yes, that would be good. Currently I resort to either manually building specific projects on the command line (tedious!) and then running specific tests (with `dotnet.testWindow.disableBuildOnRun`), or at times I have been driven back to other IDEs that support incremental dotnet builds. Iterating on tests in C# Devkit is just too slow.
